# Импортируем модуль

In [3]:
from movielens_analysis import Links, Movies, Ratings, Tags

# Загружаем данные

In [4]:
links = Links(lines=100)
tags = Tags(lines=1000)
movies = Movies(lines=1000)
ratings = Ratings(lines=1000)
movies_ratings = ratings.Movies(ratings.data, movies.data)
users = ratings.Users(ratings.data, movies.data)

links.read_csv()
tags.read_csv()
movies.read_csv()
ratings.read_csv()

Ошибка при запросе к IMDb для фильма 1: HTTPSConnectionPool(host='www.imdb.com', port=443): Max retries exceeded with url: /title/tt0114709/ (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x00000180CD859450>, 'Connection to www.imdb.com timed out. (connect timeout=5)'))


[{'userId': '1', 'movieId': '1', 'rating': '4.0', 'timestamp': '964982703'},
 {'userId': '1', 'movieId': '3', 'rating': '4.0', 'timestamp': '964981247'},
 {'userId': '1', 'movieId': '6', 'rating': '4.0', 'timestamp': '964982224'},
 {'userId': '1', 'movieId': '47', 'rating': '5.0', 'timestamp': '964983815'},
 {'userId': '1', 'movieId': '50', 'rating': '5.0', 'timestamp': '964982931'},
 {'userId': '1', 'movieId': '70', 'rating': '3.0', 'timestamp': '964982400'},
 {'userId': '1', 'movieId': '101', 'rating': '5.0', 'timestamp': '964980868'},
 {'userId': '1', 'movieId': '110', 'rating': '4.0', 'timestamp': '964982176'},
 {'userId': '1', 'movieId': '151', 'rating': '5.0', 'timestamp': '964984041'},
 {'userId': '1', 'movieId': '157', 'rating': '5.0', 'timestamp': '964984100'},
 {'userId': '1', 'movieId': '163', 'rating': '5.0', 'timestamp': '964983650'},
 {'userId': '1', 'movieId': '216', 'rating': '5.0', 'timestamp': '964981208'},
 {'userId': '1', 'movieId': '223', 'rating': '3.0', 'timestam

# Выводим результаты работы методов

## Методы класса Links

In [5]:
links.top_directors(10)

{'Martin Scorsese': 2,
 'Robert Rodriguez': 2,
 'Joe Johnston': 1,
 'Howard Deutch': 1,
 'Forest Whitaker': 1,
 'Charles Shyer': 1,
 'Michael Mann': 1,
 'Sydney Pollack': 1,
 'Peter Hewitt': 1,
 'Peter Hyams': 1}

In [6]:
%%timeit
links.top_directors(10)

60 μs ± 3.98 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода top_directors() можно получить топ режиссеров по количеству снятых картин. Исходя из работы метода выше можно увидеть, что режиссеры Martin Scorsese и Robert Rodriguez сняли по 2 фильма

In [7]:
links.most_expensive(10)

{'Остров головорезов': 98000000,
 'Храброе сердце': 72000000,
 'Денежный поезд': 68000000,
 'Джуманджи': 65000000,
 'Американский президент': 62000000,
 'Схватка': 60000000,
 'Золотой глаз': 60000000,
 'Сабрина': 58000000,
 'Покахонтас': 55000000,
 'Казино': 52000000}

In [8]:
%%timeit
links.most_expensive(10)

36.4 μs ± 1.29 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода most_expensive() можно получить топ фильмов по бюджету. Исходя из работы метода выше можно увидеть, что у фильма "Остров головорезов" был самый высокий бюджет - $98.000.000

In [53]:
links.most_total_collection(10)

{'Золотой глаз': 352194034,
 'Покахонтас': 346079773,
 'Семь': 328886817,
 'Джуманджи': 262821940,
 'Бэйб: Четвероногий малыш': 254134910,
 'Храброе сердце': 213216216,
 'Эйс Вентура 2: Когда зовёт природа': 212385533,
 'Схватка': 187436818,
 'Мосты округа Мэдисон': 182016617,
 'Опасные умы': 179519401}

In [10]:
%%timeit
links.most_total_collection(10)

38.7 μs ± 5.22 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода most_total_collection() можно получить топ фильмов по сборам. Исходя из работы метода выше можно увидеть, что у фильма "История игрушек" была самая высокие сборы - $394.436.586

In [11]:
links.most_profitable(10)

{'Семь': 295886817,
 'Золотой глаз': 292194034,
 'Покахонтас': 291079773,
 'Бэйб: Четвероногий малыш': 224134910,
 'Джуманджи': 197821940,
 'Эйс Вентура 2: Когда зовёт природа': 182385533,
 'Мосты округа Мэдисон': 158016617,
 'Опасные умы': 156519401,
 'Храброе сердце': 141216216,
 '12 обезьян': 139839459}

In [12]:
%%timeit
links.most_profitable(10)

60.2 μs ± 1.61 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода most_profitable() можно получить топ фильмов по прибыли(мировой доход - бюджет на производство). К примеру исходя из работы метода выше можно увидеть, что у фильма 'История игрушек' самая большая прибыль равная - $364.436.586

In [13]:
links.longest(10)

{'Никсон': 192,
 'Казино': 178,
 'Храброе сердце': 178,
 'Отверженные': 175,
 'Схватка': 170,
 'Опус мистера Холланда': 144,
 'Разум и чувства': 136,
 'Мосты округа Мэдисон': 135,
 'Наемные убийцы': 133,
 'Золотой глаз': 130}

In [14]:
%%timeit
links.longest(10)

29.5 μs ± 661 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода longest() можно получить топ фильмов по продолжительности. Исходя из работы метода выше можно увидеть, что фильм 'Никсон' самый продолжительный в выборке - 192мин

In [15]:
links.top_cost_per_minute(10)


{'Остров головорезов': 790322.58,
 'Покахонтас': 679012.35,
 'Джуманджи': 625000.0,
 'Денежный поезд': 618181.82,
 'Чeстная игра': 549450.55,
 'Американский президент': 543859.65,
 'Индеец в шкафу': 468750.0,
 'Сломанная стрела': 462962.96,
 'Золотой глаз': 461538.46,
 'Сабрина': 456692.91}

In [16]:
%%timeit
links.top_cost_per_minute(10)

165 μs ± 4.99 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода top_cost_per_minute() можно получить результат стоимости 1 минуты фильма. Исходя из работы метода выше можно увидеть, что фильм 'Остров головорезов' по результатам вычислений имеет наибольшее значение - $790.322,58

In [17]:
links.most_ratings(10)

{'Семь': 8.6,
 'Подозрительные лица': 8.5,
 'Схватка': 8.3,
 'Храброе сердце': 8.3,
 'Казино': 8.2,
 'Таксист': 8.2,
 'Ненависть': 8.1,
 '12 обезьян': 8.0,
 'Почтальон': 7.8,
 'Разум и чувства': 7.7}

In [18]:
%%timeit
links.most_ratings(10)

27.5 μs ± 1.16 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### В ходе выполнения метода most_ratings() можно получить топ фильмов по рейтингу. Исходя из работы метода выше можно увидеть, что фильм 'Семь' самый рейтинговый - 8.6

In [54]:
links.most_popularity(30)

{'Семь': 129,
 'Схватка': 261,
 'Таксист': 286,
 'Подозрительные лица': 302,
 'Храброе сердце': 402,
 'От заката до рассвета': 509,
 'Казино': 734,
 'Достать коротышку': 761,
 'Джуманджи': 960,
 'Разум и чувства': 965,
 'Бестолковые': 1061,
 'Счастливчик Гилмор': 1235,
 '12 обезьян': 1257,
 'Золотой глаз': 1264,
 'Пятница': 1420,
 'Мертвец идет': 1463,
 'Покидая Лас-Вегас': 1711,
 'Покахонтас': 2061,
 'Ненависть': 2253,
 'Умереть во имя': 2420,
 'Мосты округа Мэдисон': 2427,
 'Смертельная битва': 2544,
 'Сейчас и тогда': 2783,
 'Опус мистера Холланда': 2987,
 'Доводы рассудка': 3069,
 'Не грози южному централу, попивая сок у себя в квартале': 3661,
 'Эйс Вентура 2: Когда зовёт природа': 3673,
 'Бэйб: Четвероногий малыш': 3977,
 'Пудра': 4072,
 'Четыре комнаты': 4268}

In [20]:
%%timeit
links.most_popularity(10)

16.4 μs ± 180 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


#### В ходе выполнения метода most_popularity() можно получить топ фильмов по поулярности. Исходя из работы метода выше можно увидеть, что фильм 'Семь' самый популярный - занимает 129 строчку

## Методы класса Tags

In [21]:
tags.most_words(5)

{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'Oscar (Best Music - Original Score)': 6,
 'Everything you want is here': 5,
 'based on a true story': 5}

In [22]:
%%timeit
tags.most_words(5)

341 μs ± 18 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода most_words() можно получить топ тегов по количеству слов. Из работы метода видно, что из выборки наибольшее количество слов в теге равняется 16

In [23]:
tags.longest(5)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'audience intelligence underestimated',
 'Oscar (Best Music - Original Score)',
 'assassin-in-training (scene)']

In [24]:
%%timeit
tags.longest(5)

70.5 μs ± 1.17 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### Метод longest() возвращает список самых длинных тегов по количеству символов(включая пробел).

In [55]:
tags.most_words_and_longest(10)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'Oscar (Best Music - Original Score)',
 'Everything you want is here',
 'Guardians of the Galaxy']

In [26]:
%%timeit
tags.most_words_and_longest(5)

452 μs ± 18.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### Метод most_words_and_longest() возвращает пересечение между тегами с наибольшим количеством слов внутри и тегами с наибольшей длиной по количеству символов

In [27]:
tags.most_popular(5)

{'funny': 15,
 'sci-fi': 14,
 'twist ending': 12,
 'dark comedy': 12,
 'atmospheric': 10}

In [28]:
%%timeit
tags.most_popular(5)

278 μs ± 6.72 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### Метод most_popular() возвращает самые популярные теги. Исходя из работы метода видим, что наиболее популярным тегом является тег 'funny', встречабщийся 15 раз

In [29]:
tags.tags_with('funny')

['funny', 'stupid but funny', 'very funny']

In [30]:
%%timeit
tags.tags_with('funny')

139 μs ± 4.61 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### Метод tags_with() возвращает уникальные теги, содержащие слово. Исходя из работы метода мы видим, что есть 3 уникальных тега, содержащих слово 'funny'

## Методы класса Movies

In [31]:
movies.dist_by_release()

{1995: 224,
 1994: 184,
 1996: 181,
 1993: 101,
 1992: 23,
 1990: 15,
 1991: 15,
 1989: 14,
 1986: 9,
 1982: 8,
 1940: 8,
 1957: 8,
 1987: 8,
 1980: 8,
 1981: 7,
 1988: 7,
 1979: 7,
 1955: 6,
 1959: 6,
 1968: 6,
 1997: 6,
 1939: 6,
 1985: 6,
 1967: 5,
 1965: 5,
 1951: 5,
 1958: 5,
 1944: 5,
 1941: 5,
 1975: 5,
 1971: 5,
 1984: 5,
 1964: 4,
 1973: 4,
 1954: 4,
 1934: 4,
 1960: 4,
 1963: 4,
 1950: 4,
 1974: 4,
 1983: 4,
 1977: 3,
 1937: 3,
 1972: 3,
 1952: 3,
 1961: 3,
 1953: 3,
 1946: 3,
 1938: 3,
 1956: 3,
 1962: 3,
 1976: 2,
 1969: 2,
 1970: 2,
 1942: 2,
 1945: 2,
 1947: 2,
 1935: 2,
 1936: 2,
 1949: 2,
 1978: 2,
 1943: 1,
 1932: 1,
 1966: 1,
 1948: 1,
 1933: 1,
 1931: 1}

In [32]:
%%timeit
movies.dist_by_release()

531 μs ± 17.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### Метод dist_by_release() возвращает количество фильмов по годам. В результате работs метода мы видим, что исходя из выборки наибольшее количество фильмов произведено 1995, их 224

In [33]:
movies.dist_by_genres()

{'Drama': 507,
 'Comedy': 365,
 'Romance': 208,
 'Thriller': 179,
 'Action': 158,
 'Adventure': 126,
 'Crime': 122,
 'Children': 100,
 'Fantasy': 69,
 'Sci-Fi': 69,
 'Mystery': 58,
 'Musical': 53,
 'Horror': 51,
 'War': 48,
 'Animation': 37,
 'Documentary': 25,
 'Western': 23,
 'Film-Noir': 18,
 '(no genres listed)': 3}

In [34]:
%%timeit
movies.dist_by_genres()

1.5 ms ± 39.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### Метод dist_by_genres() возвращает количество по жанров. В результате работs метода мы видим, что исходя из выборки наибольшее количество жанров это 'Drama', их 507

In [35]:
movies.most_genres(5)

{'Drama': 507, 'Comedy': 365, 'Romance': 208, 'Thriller': 179, 'Action': 158}

In [36]:
%%timeit
movies.most_genres(5)

1.5 ms ± 16 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### Метод most_genres() возвращает топ жанров. В результате работы метода мы видим, что исходя из выборки наименьшее количество жанров из топ 5 это 'Action', их 158

## Методы класса Ratings.Movies

In [37]:
movies_ratings.dist_by_year()

{2015: 29,
 2011: 39,
 2007: 1,
 2006: 4,
 2005: 121,
 2001: 70,
 2000: 296,
 1999: 82,
 1996: 358}

In [38]:
%%timeit
movies_ratings.dist_by_year()

1.06 ms ± 20 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода dist_by_year() можно получить количество фильмов по годам. Из работы метода видно распределение фильмов по годам

In [39]:
movies_ratings.dist_by_rating()

{0.5: 24,
 1.0: 39,
 1.5: 11,
 2.0: 57,
 2.5: 7,
 3.0: 253,
 3.5: 17,
 4.0: 292,
 4.5: 33,
 5.0: 267}

In [40]:
%%timeit
movies_ratings.dist_by_rating()

489 μs ± 21.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода dist_by_rating() можно получить количсетво фильмов по оценкам. Из работы метода видно распределение фильмов по оценкам

In [56]:
movies_ratings.top_by_num_of_ratings(20)

{'Usual Suspects The': 4,
 'Pulp Fiction': 4,
 'Fugitive The': 4,
 "Schindler's List": 4,
 'Batman': 4,
 'Silence of the Lambs The': 4,
 'Fargo': 4,
 'Aladdin': 4,
 'Beauty and the Beast': 4,
 'Toy Story': 3,
 'Seven (a.k.a. Se7en)': 3,
 'Braveheart': 3,
 'Star Wars: Episode IV - A New Hope': 3,
 'Stargate': 3,
 'Tommy Boy': 3,
 'Clear and Present Danger': 3,
 'Forrest Gump': 3,
 'Mask The': 3,
 'Jurassic Park': 3,
 'Tombstone': 3}

In [42]:
%%timeit
movies_ratings.top_by_num_of_ratings(10)

129 ms ± 4.82 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### В ходе выполнения метода top_by_num_of_ratings() можно получить количсетво рейтингов по фильмам. Из работы метода видно, как выглядит топ-10 по рейтингам

In [43]:
movies_ratings.top_by_ratings(5, 'median')

{'Grumpier Old Men': 5.0,
 'Bottle Rocket': 5.0,
 'Rob Roy': 5.0,
 'Canadian Bacon': 5.0,
 'Desperado': 5.0}

In [44]:
%%timeit
movies_ratings.top_by_ratings(5, 'median')

75.2 ms ± 2.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### В ходе выполнения метода top_by_ratings() можно получить топ фильмов по рейтингу в зависимости от определенной метрики. К примеру, топ 5 по медианному рейтингу 

In [45]:
movies_ratings.top_controversial(5)

{'My Fair Lady': 10.12,
 'Courage Under Fire': 6.12,
 "Schindler's List": 4.56,
 'Dazed and Confused': 4.5,
 'Circle of Friends': 4.5}

In [46]:
%%timeit
movies_ratings.top_controversial(5)

131 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### В ходе выполнения метода top_controversial() можно получить топ лучших фильмов по дисперсии рейтингов. К примеру исходя из работы метода можем увидеть, что фильм 'My Fair Lady' имеет значение 10.12

## Методы класса Users(Movies)

In [47]:
users.user_by_num_of_ratings(5)

{'6': 314, '1': 232, '4': 216, '7': 126, '5': 44}

In [48]:
%%timeit
users.user_by_num_of_ratings(5)

264 μs ± 11.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода user_by_num_of_ratings() можно получить распределение пользователей по количеству выставленных ими оценок. Исходя из работы метода можно увидеть, что пользователь с userId равным 6 поставил 314 оценок

In [49]:
users.user_by_ratings(5)

{'2': 70350.28, '3': 7058.38, '7': 4790.09, '4': 1982.13, '1': 1854.6}

In [50]:
%%timeit
users.user_by_ratings(5)

485 μs ± 21.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода user_by_ratings() можно получить распределение пользователей по средним или медианным оценкам, выставленным ими. Исходя из работы метода можно увидеть, что пользователь с userId равным 2 имеет среднюю медианную оценку равную 70350.28

In [51]:
users.user_controversial(5)

{'3': 4.37, '4': 1.73, '7': 1.67, '5': 0.98, '6': 0.72}

In [52]:
%%timeit
users.user_controversial(5)

726 μs ± 6.03 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### В ходе выполнения метода user_controversial() можно получить пользователей с наибольшей разницей в рейтингах. Исходя из работы метода можно увидеть, что пользователь с userId равным 3 имеет наибольшую разницу в оценках, равную 4.37